archive/imo/imo2019_q2.lean

```diff
@@ -0,0 +1,576 @@
+    sphere.second_inter_vsub_mem_affine_span _ _ _,
+  have hB' : cfg.A₂ ∈ line[ℝ, cfg.triangle_ABC.points 1, cfg.B₁],
+  { rw ←h, exact sphere.second_inter_vsub_mem_affine_span _ _ _ },
+  exact (sbtw_of_sbtw_of_sbtw_of_mem_affine_span_pair dec_trivial hA hB hA' hB').symm.not_rotate
+    cfg.sbtw_A_A₁_A₂.wbtw
+end
+
+lemma wbtw_B_Q_B₂ : wbtw ℝ cfg.B cfg.Q cfg.B₂ := cfg.sbtw_B_B₁_B₂.wbtw.trans_left cfg.wbtw_B_Q_B₁
+
+section oriented
+
+variables [module.oriented ℝ V (fin 2)]
+
+lemma two_zsmul_oangle_QPA₂_eq_two_zsmul_oangle_BAA₂ :
+  (2 : ℤ) • ∡ cfg.Q cfg.P cfg.A₂ = (2 : ℤ) • ∡ cfg.B cfg.A cfg.A₂ :=
+begin
+  refine two_zsmul_oangle_of_parallel cfg.QP_parallel_BA _,
+  convert affine_subspace.parallel.refl _ using 1,
+  rw [cfg.collinear_PAA₁A₂.affine_span_eq_of_ne
+        (set.mem_insert_of_mem _ (set.mem_insert_of_mem _ (set.mem_insert_of_mem _
+          (set.mem_singleton _))))
+        (set.mem_insert_of_mem _ (set.mem_insert _ _)) cfg.A₂_ne_A,
+      cfg.collinear_PAA₁A₂.affine_span_eq_of_ne
+        (set.mem_insert_of_mem _ (set.mem_insert_of_mem _ (set.mem_insert_of_mem _
+          (set.mem_singleton _))))
+        (set.mem_insert _ _) cfg.A₂_ne_P]
+end
+
+end oriented
+
+lemma not_collinear_QPA₂ : ¬ collinear ℝ ({cfg.Q, cfg.P, cfg.A₂} : set Pt) :=
+begin
+  haveI := some_orientation V,
+  rw [collinear_iff_of_two_zsmul_oangle_eq cfg.two_zsmul_oangle_QPA₂_eq_two_zsmul_oangle_BAA₂,
+      ←affine_independent_iff_not_collinear_set],
+  have h : cospherical ({cfg.B, cfg.A, cfg.A₂} : set Pt),
+  { refine cfg.triangle_ABC.circumsphere.cospherical.subset _,
+    simp [set.insert_subset, cfg.A_mem_circumsphere, cfg.B_mem_circumsphere,
+          cfg.A₂_mem_circumsphere] },
+  exact h.affine_independent_of_ne cfg.A_ne_B.symm cfg.A₂_ne_B.symm cfg.A₂_ne_A.symm
+end
+
+lemma Q₁_ne_A₂ : cfg.Q₁ ≠ cfg.A₂ :=
+begin
+  intro h,
+  have h₁ := cfg.sbtw_Q_A₁_Q₁,
+  rw h at h₁,
+  refine cfg.not_collinear_QPA₂ _,
+  have hA₂ := cfg.sbtw_A_A₁_A₂.right_mem_affine_span,
+  have hA₂A₁ : line[ℝ, cfg.A₂, cfg.A₁] ≤ line[ℝ, cfg.A, cfg.A₁] :=
+    affine_span_pair_le_of_left_mem hA₂,
+  have hQ : cfg.Q ∈ line[ℝ, cfg.A, cfg.A₁],
+  { rw affine_subspace.le_def' at hA₂A₁,
+    exact hA₂A₁ _ h₁.left_mem_affine_span },
+  exact collinear_triple_of_mem_affine_span_pair hQ cfg.wbtw_A_P_A₁.mem_affine_span hA₂
+end
+
+lemma affine_independent_QPA₂ : affine_independent ℝ ![cfg.Q, cfg.P, cfg.A₂] :=
+affine_independent_iff_not_collinear_set.2 cfg.not_collinear_QPA₂
+
+lemma affine_independent_PQB₂ : affine_independent ℝ ![cfg.P, cfg.Q, cfg.B₂] :=
+by { rw ←symm_A₂, exact cfg.symm.affine_independent_QPA₂ }
+
+/-- `QPA₂` as a `triangle`. -/
+def triangle_QPA₂ : triangle ℝ Pt := ⟨_, cfg.affine_independent_QPA₂⟩
+
+/-- `PQB₂` as a `triangle`. -/
+def triangle_PQB₂ : triangle ℝ Pt := ⟨_, cfg.affine_independent_PQB₂⟩
+
+lemma symm_triangle_QPA₂ : cfg.symm.triangle_QPA₂ = cfg.triangle_PQB₂ :=
+by { simp_rw [triangle_PQB₂, ←symm_A₂], ext i, fin_cases i; refl }
+
```

Comment:
Is it worth somehow flagging in this proof that the previous few hundred lines have been basically proving that a bunch of points in the diagram are distinct, something which would presumably typically not be mentioned at all by an IMO student? For example a comment on line 125 or so saying "the next 300 or so lines are proving things such as `Q₁ ≠ A₂` which would be typically not mentioned in a paper proof"? Or perhaps indicating something like "the proof proper starts here" because this is in some sense where the ideas start.

Reply:
I added a remark to the module doc string about how it's mostly proving nondegeneracy conditions.

Reply:
It would be great to add some `/-! ### Subsection -/` headings to make this even clearer.

Reply:
I've added a series of such headings.